applications/nrf5340_audio/src/modules/sd_card_playback.c

```diff
@@ -0,0 +1,411 @@
+static k_tid_t sd_card_playback_thread_id;
+
+/* Playback */
+static bool sd_card_playback_active;
+static char *playback_file_name;
+static enum audio_formats playback_file_format;
+static uint16_t pcm_frame_size;
+static char playback_file_path[MAX_PATH_LEN] = "";
+static struct lc3_binary_hdr_t lc3_file_header;
+static struct wav_playback_config wav_playback_cfg = {.frame_duration_ms = 10, .bit_depth = 16};
+static struct lc3_playback_config lc3_playback_cfg = {.lc3_frame_length_size = FR_LEN_INFO_SIZE};
+
+static struct fs_file_t f_seg_read_entry;
+
+static int sd_card_playback_buffer_set(uint8_t *buf, size_t *size)
+{
+	int ret;
+	uint16_t write_size;
+
+	ret = k_mutex_lock(&mtx_ringbuf, K_MSEC(1));
+	if (ret < 0) {
+		LOG_ERR("Unable to take mutex. Ret: %d", ret);
+		return ret;
+	}
+	write_size = ring_buf_get(&m_ringbuf_sound_data_lc3, buf, *size);
+	ret = k_mutex_unlock(&mtx_ringbuf);
+	if (ret < 0) {
+		LOG_ERR("Mutex unlock err: %d", ret);
+		return ret;
+	}
+	if (ring_buf_space_get(&m_ringbuf_sound_data_lc3) >= pcm_frame_size) {
+		k_sem_give(&m_sem_load_from_buf_lc3);
+	}
+
+	*size = write_size;
+
+	return 0;
+}
+
+static int sd_card_playback_buffer_to_ringbuffer(uint8_t *buffer, size_t numbytes)
+{
+	int ret;
+	uint8_t *buf_ptr;
+
+	ret = k_mutex_lock(&mtx_ringbuf, K_MSEC(1));
+	if (ret < 0) {
+		LOG_ERR("Unable to take mutex. Ret: %d", ret);
+		return ret;
+	}
+	numbytes = ring_buf_put_claim(&m_ringbuf_sound_data_lc3, &buf_ptr, numbytes);
+	memcpy(buf_ptr, buffer, numbytes);
+	ring_buf_put_finish(&m_ringbuf_sound_data_lc3, numbytes);
+	ret = k_mutex_unlock(&mtx_ringbuf);
+	if (ret < 0) {
+		LOG_ERR("Mutex unlock err: %d", ret);
+		return ret;
+	}
+	return numbytes;
+}
+
+static int sd_card_playback_load_ringbuf(uint8_t *pcm_frame, size_t write_size)
+{
+	int ret;
+	ret = k_sem_take(&m_sem_load_from_buf_lc3, K_MSEC(30));
+	if (ret < 0) {
+		LOG_ERR("Sem take err: %d. Skipping frame", ret);
+		return ret;
+	}
+	sd_card_playback_buffer_to_ringbuffer(pcm_frame, write_size);
+	return 0;
+}
+
+static int sd_card_playback_play_lc3(void)
+{
+	int ret;
+	size_t lc3_file_header_size = sizeof(lc3_file_header);
+	uint16_t pcm_mono_write_size;
+
+	ret = sd_card_open(playback_file_name, &f_seg_read_entry);
+	if (ret < 0) {
+		LOG_ERR("Open SD card file err: %d", ret);
+		return ret;
+	}
+
+	/* Read the header */
+	ret = sd_card_read((char *)&lc3_file_header, &lc3_file_header_size, &f_seg_read_entry);
+	if (ret < 0) {
+		LOG_ERR("Read SD card file err: %d", ret);
+		return ret;
+	}
+
+	pcm_frame_size = 2 * lc3_file_header.sample_rate * lc3_file_header.frame_ms / 1000;
+	lc3_playback_cfg.lc3_frames_num =
+		2 * ((lc3_file_header.signal_len_msb << 16) + lc3_file_header.signal_len_lsb) /
+		pcm_frame_size;
+
+	uint8_t pcm_mono_frame[pcm_frame_size];
+
+	sd_card_playback_active = true;
+	for (uint32_t i = 0; i < lc3_playback_cfg.lc3_frames_num; i++) {
+
+		uint8_t lc3_frame[lc3_playback_cfg.lc3_frame_length];
```

Comment:
There might be a better way of doing it, but I am using it as a uint16 some places. If I have a char buffer, then I think I have to struggle a bit because I need to fetch the info from two indices in the buffer

Reply:
That should not matter in this scope actually. If you have a look at `sizeof` for both `char` and `uint8_t`, you'll see that they're both 1. So any manipulation you have to do to access the array as `uint16_t` will be the same for both (TLDR `char` and `uint8_t` are the same). Since the two functions that are using this array both take the pointer as a `char *` it's safe to define the array as a `char` - if any of the functions need to access it as `uint16_t` they will do the casting themselves (y)